neqo-transport/src/connection/tests/stream.rs

```diff
@@ -111,6 +114,179 @@ fn transfer() {
+    let mut streams = Vec::<StreamId>::new();
+    for sendorder in order_of_sendorder {
+        let id = client.stream_create(StreamType::UniDi).unwrap();
+        streams.push(id);
+        ordered.push((id, *sendorder));
+        // must be set before sendorder
+        client.streams.set_fairness(id, true).ok();
+        client.streams.set_sendorder(id, *sendorder).ok();
+    }
+    // Write some data to all the streams
+    for stream_id in streams {
+        client.stream_send(stream_id, &[6; 100]).unwrap();
+    }
+
+    // Sending this much takes a few datagrams.
+    // Note: this test uses an RTT of 0 which simplifies things (no pacing)
+    let mut datagrams = Vec::new();
+    let mut out = client.process_output(now());
+    while let Some(d) = out.dgram() {
+        datagrams.push(d);
+        out = client.process_output(now());
+    }
+    assert_eq!(*client.state(), State::Confirmed);
+
+    qdebug!("---- server receives");
+    for (_, d) in datagrams.into_iter().enumerate() {
+        let out = server.process(Some(d), now());
+        qdebug!("Output={:0x?}", out.as_dgram_ref());
+    }
+    assert_eq!(*server.state(), State::Confirmed);
+
+    let stream_ids = server.events().filter_map(|evt| match evt {
+        ConnectionEvent::RecvStreamReadable { stream_id, .. } => Some(stream_id),
+        _ => None,
+    }).enumerate().map(|(a, b)| (b, a)).collect::<HashMap<_, _>>();
+
+    // streams should arrive in priority order, not order of creation, if sendorder prioritization
+    // is working correctly
+
+    // 'ordered' has the send order currently.  Re-sort it by sendorder, but
+    // if two items from the same sendorder exist, secondarily sort by the ordering in
+    // the stream_ids vector (HashMap<StreamId, index: usize>)
+    ordered.sort_unstable_by_key(|(stream_id, sendorder)| (StreamOrder { sendorder: *sendorder}, stream_ids[stream_id]));
+    // make sure everything now is in the same order, since we modified the order of
+    // same-sendorder items to match the ordering of those we saw in reception
+    for (i, (stream_id, _sendorder)) in ordered.iter().enumerate() {
+        assert_eq!(i, stream_ids[stream_id]);
+    }
+}
+
+#[test]
+fn sendorder_0() {
+    sendorder_test(&[None, Some(1), Some(2), Some(3)]);
+}
+#[test]
+fn sendorder_1() {
+    sendorder_test(&[Some(3), Some(2), Some(1), None]);
+}
+#[test]
+fn sendorder_2() {
+    sendorder_test(&[Some(3), None, Some(2), Some(1)]);
+}
+#[test]
+fn sendorder_3() {
+    sendorder_test(&[Some(1), Some(2), None, Some(3)]);
+}
+#[test]
+fn sendorder_4() {
+    sendorder_test(&[Some(1), Some(2), Some(1), None, Some(3), Some(1), Some(3), None]);
+}
+
+
+// Tests stream sendorder priorization
+// Converts Vecs of u64's into StreamIds
+fn fairness_test<S, R, SI>(source: S, number_iterates: usize, truncate_to: usize, result_array: R)
+where
+    S: IntoIterator<Item = SI>,
+    R: IntoIterator<Item = SI> + std::fmt::Debug,
+    SI: Into<StreamId>, Vec<u64>: PartialEq<R>
```

Comment:
As a general rule, you shouldn't constrain the caller by requiring the same type for items in both iterators.

That means something like this (or separate SI and RI type parameters):

```rust
where
S: IntoIterator,
S::Item: Into<StreamId>,
R: IntoIterator,
R::Item: Into<StreamId>,
Vec<u64>: PartialEq<R>,
```

However, it looks like the last constraint is only one you really need.  That is, you only use PartialEq.  R doesn't need to implement IntoIterator because you don't use that.

If I were designing an API, I probably wouldn't restrict the caller that way.  You would instead use `Iterator::eq` to compare items from the expected outcome with what really happened.  But that probably isn't worth it in this case, especially since that would change what is printed when the assertion fails, and not in a good way.